drivers/can/can_common.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <can.h>
+#include <kernel.h>
+
+#define LOG_LEVEL CONFIG_CAN_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(can_driver);
+
+#define WORK_BUF_COUNT_IS_POWER_OF_2 !(CONFIG_CAN_WORKQ_FRAMES_BUF_CNT & \
+					(CONFIG_CAN_WORKQ_FRAMES_BUF_CNT - 1))
+
+#define WORK_BUF_MOD_MASK (CONFIG_CAN_WORKQ_FRAMES_BUF_CNT - 1)
+
+#if WORK_BUF_COUNT_IS_POWER_OF_2
+#define WORK_BUF_MOD_SIZE(x) ((x) & WORK_BUF_MOD_MASK)
+#else
+#define WORK_BUF_MOD_SIZE(x) ((x) % CONFIG_CAN_WORKQ_FRAMES_BUF_CNT)
+#endif
+
+#define WORK_BUF_FULL 0xFFFF
+
+static void can_msgq_put(struct zcan_frame *frame, void *arg)
+{
+	struct k_msgq *msgq = (struct k_msgq *)arg;
+	int ret;
+
+	__ASSERT_NO_MSG(msgq);
+
+	ret = k_msgq_put(msgq, frame, K_NO_WAIT);
+	if (ret) {
+		LOG_ERR("Msgq %p overflowed. Frame ID: 0x%x", arg,
+			frame->id_type == CAN_STANDARD_IDENTIFIER ?
+				frame->std_id : frame->ext_id);
+	}
+}
+
+int z_impl_can_attach_msgq(struct device *dev, struct k_msgq *msg_q,
+			   const struct zcan_filter *filter)
+{
+	const struct can_driver_api *api = dev->driver_api;
+
+	return api->attach_isr(dev, can_msgq_put, msg_q, filter);
+}
+
+static inline void can_work_buffer_init(struct can_frame_buffer *buffer)
+{
+	buffer->head = 0;
+	buffer->tail = 0;
+}
+
+static inline int can_work_buffer_put(struct zcan_frame *frame,
+				      struct can_frame_buffer *buffer)
+{
+	u16_t next_head = WORK_BUF_MOD_SIZE(buffer->head + 1);
+
+	if (buffer->head == WORK_BUF_FULL) {
+		return -1;
```

Comment:
Good point. I could also use void and throw the error log here.

Reply:
-EIO seems not really to be ideal here because it is not an IO error. I can't find a suitable error number like buffer full. Since this is defined static and only used at one location I leave it as -1 here